backend/src/Notifo.Domain.Integrations/Discord/DiscordBotClientPool.cs

```diff
@@ -0,0 +1,36 @@
+﻿// =====================================================
+//  Notifo.io
+// ==========================================================================
+//  Copyright (c) Sebastian Stehle
+//  Author of the file: Artur Nowak
+//  All rights reserved. Licensed under the MIT license.
+// ==========================================================================
+
+using Discord;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Notifo.Domain.Integrations.Discord;
+public class DiscordBotClientPool : CachePool<IDiscordClient>
+{
+    public DiscordBotClientPool(IMemoryCache memoryCache)
+        : base(memoryCache)
+    {
+    }
+
+    public IDiscordClient GetDiscordClient(string botToken, CancellationToken ct)
+    {
+        var cacheKey = $"{nameof(IDiscordClient)}_{botToken}";
+
+        var found = GetOrCreate(cacheKey, TimeSpan.FromMinutes(5), () =>
+        {
+            var client = new DiscordClient();
+
+            client.LoginAsync(TokenType.Bot, botToken).Wait(ct);
```

Comment:
Never, ever use Wait(). We can make an async version of the GetOrCreate method.

All methods of the client accept a RequestOption with cancellationToken. We should always forward that if possible.

Reply:
Sure, thank you so much!

Sadly the `LoginAsync` method from `DiscordRestClient` (or rather `BaseDiscordClient`) from Discord.NET is a wrapper over the internal `ApiClient.LoginAsync` method (this one accepts the RequestOptions parameter, which has a CancellationToken property). But the base function calls it with the default RequestOptions parameter.

Therefore, the problem is that the `LoginAsync` function doesn't expose the `RequestOptions` parameter (while other operations like e.g. SendMessage do expose it).

Technically  I could write some kind of a workaround (e.g. `WaitAsync`), but even if the cancellation token fires and WaitAsync completes, the LoginAsync Task will still get executed on that another thread. On the other hand, interrupting the Thread with it seems like overengineering and can lead to some integrity issues.

What do you think? Maybe I should just forget about the cancellation token in that specific method?